sdk/eventhub/event-hubs/src/eventHubReceiver.ts

```diff
@@ -200,214 +165,233 @@ export class EventHubReceiver extends LinkEntity {
-              "retryable, we let the user know about it by calling the user's error handler.",
-            this._context.connectionId
-          );
-          this._onError!(ehError);
-        }
-      }
-    };
+    if (this.receiverRuntimeMetricEnabled && data) {
+      this.runtimeInfo.lastEnqueuedSequenceNumber = data.lastSequenceNumber;
+      this.runtimeInfo.lastEnqueuedTimeUtc = data.lastEnqueuedTime;
+      this.runtimeInfo.lastEnqueuedOffset = data.lastEnqueuedOffset;
+      this.runtimeInfo.retrievalTime = data.retrievalTime;
+      log.receiver(
+        "[%s] RuntimeInfo of Receiver '%s' is %O",
+        this._context.connectionId,
+        this.name,
+        this.runtimeInfo
+      );
+    }
 
-    this._onAmqpClose = async (context: EventContext) => {
-      const receiverError = context.receiver && context.receiver.error;
-      const receiver = this._receiver || context.receiver!;
-      if (receiverError) {
-        log.error(
-          "[%s] 'receiver_close' event occurred for receiver '%s' with address '%s'. " +
-            "The associated error is: %O",
-          this._context.connectionId,
-          this.name,
-          this.address,
-          receiverError
-        );
-      }
-      if (receiver && !receiver.isItselfClosed()) {
-        if (!this.isConnecting) {
-          log.error(
-            "[%s] 'receiver_close' event occurred on the receiver '%s' with address '%s' " +
-              "and the sdk did not initiate this. The receiver is not reconnecting. Hence, calling " +
-              "detached from the _onAmqpClose() handler.",
-            this._context.connectionId,
-            this.name,
-            this.address
-          );
-          await this.onDetached(receiverError);
-        } else {
-          log.error(
-            "[%s] 'receiver_close' event occurred on the receiver '%s' with address '%s' " +
-              "and the sdk did not initate this. Moreover the receiver is already re-connecting. " +
-              "Hence not calling detached from the _onAmqpClose() handler.",
-            this._context.connectionId,
-            this.name,
-            this.address
-          );
-        }
-      } else {
-        if (this._abortSignal) {
-          this._abortSignal.removeEventListener("abort", this._onAbort);
-        }
-        log.error(
-          "[%s] 'receiver_close' event occurred on the receiver '%s' with address '%s' " +
-            "because the sdk initiated it. Hence not calling detached from the _onAmqpClose" +
-            "() handler.",
-          this._context.connectionId,
-          this.name,
-          this.address
-        );
-      }
-    };
+    // automatically add credit if there is a listener
+    if (this._onMessage && !this._usingInternalQueue) {
+      const existingCredits = this._receiver ? this._receiver.credit : 0;
+      this._addCredit(Math.max(this.prefetchCount - existingCredits, 0));
```

Comment:
Say there is a `receiveBatch(10)` call.
We get 1 message.
`existingCredits` would be 9, `prefetchCount` would be 10` and the above code will add 1 credit.
Why would we want to add a credit in this scenario?

Reply:
I think my original thought was to maintain credits since I started with the streaming method.

I think what I’m going to have to do is manage credits from the receiveBatch and receive methods in receiver.ts instead.

Reply:
I updated this logic so that this will only add credits if `this.prefetchCount` is not 0. Also, `this.prefetchCount` is now only set to something greater than 0 when calling the `receive` (streaming) operation, so this essentially maintains our credits as the maximum set when streaming.